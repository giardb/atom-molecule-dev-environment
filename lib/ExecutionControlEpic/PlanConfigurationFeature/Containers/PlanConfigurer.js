"use babel";
// @flow

import { connect } from "react-redux";
import PlanConfigurer from "../Presenters/PlanConfigurer";
import type { State } from "../../../GlobalSystem/types.js.flow";
import { selectConfigSchemaOfTool } from "../Selectors/PlanConfigSchemas";
import type { DevToolPlanConfigSchema, Stager } from "../Types/types.js.flow";
import {
  selectPackagesReducer,
  selectPlansSchemaReducer,
} from "../../../GlobalSystem/Selectors";
import { addPlanConfig } from "../Actions/AddPlanConfig";
import type {
  Package,
  PackageInfos,
} from "../../../ProjectSystemEpic/PackageFeature/Types/types.js.flow";
import { selectPackagesOfTool } from "../../../ProjectSystemEpic/PackageFeature/Selectors/Packages";
import { compose, withState, lifecycle, withProps } from "recompose";
import {
  FormController,
  StateDispatcher,
  convertConversionModelToConversionJobs,
  convertIn,
  convertOut,
  composeValidation,
  validateModel,
  notEmpty,
  // $FlowFixMe
} from "react-forms-state";
import { getDefault } from "../Model/PlanConfigManipulators";

export function mapStateToProps(
  state: State,
  { toolId }: { toolId: string },
): { config: DevToolPlanConfigSchema, packages: Array<Package> } {
  return {
    config: selectConfigSchemaOfTool(selectPlansSchemaReducer(state), toolId),
    packages: selectPackagesOfTool(selectPackagesReducer(state), toolId),
  };
}

export function mapDispatchToProps(
  dispatch: (action: any) => any,
  ownProps: { toolId: string, toolIconUri: string, toolName: string },
): {
  onAddPlan: (plan: {
    name: string,
    config: mixed,
    stager: Stager,
    packageInfos: PackageInfos,
  }) => void,
} {
  return {
    onAddPlan: (plan: {
      name: string,
      config: mixed,
      stager: Stager,
      packageInfos: PackageInfos,
    }) => {
      dispatch(
        addPlanConfig(
          plan.name,
          {
            name: ownProps.toolName,
            id: ownProps.toolId,
            iconUri: ownProps.toolIconUri,
          },
          plan.config,
          plan.stager,
          plan.packageInfos,
        ),
      );
    },
  };
}

export var Connecter = connect(mapStateToProps, mapDispatchToProps);

function convertsConfigSchemaToFormModel(
  value,
  props,
  configSchema,
  context = {},
) {
  if (configSchema == null) return undefined;
  switch (configSchema.type) {
    case "array":
      return {
        out: context.fieldName,
        validate: configSchema.validate,
        convertIn: array =>
          array.reduce((red, v, i) => ({ ...red, [i]: v }), {}),
        convertOut: map =>
          Object.keys(map).reduce(
            (red, i) => (map[i] != null ? red.concat(map[i]) : red),
            [],
          ),
        ...Object.keys(value || {}).reduce(
          (red, i) => ({
            ...red,
            [i]: convertsConfigSchemaToFormModel(
              value[i],
              props,
              configSchema.items,
              { fieldName: i },
            ),
          }),
          {},
        ),
      };
    case "object":
      return {
        out: context.fieldName,
        validate: configSchema.validate,
        ...Object.keys(configSchema.schemas).reduce(
          (red, key) => ({
            ...red,
            [key]: convertsConfigSchemaToFormModel(
              value[key],
              props,
              configSchema.schemas[key],
              { fieldName: key },
            ),
          }),
          {},
        ),
      };
    case "conditional":
      return {
        out: context.fieldName,
        expressionValue: convertsConfigSchemaToFormModel(
          value ? value.expressionValue : null,
          props,
          configSchema.expression,
          { fieldName: "expressionValue" },
        ),
        caseValue: convertsConfigSchemaToFormModel(
          value ? value.caseValue : null,
          props,
          configSchema.cases[value.expressionValue],
          { fieldName: "caseValue" },
        ),
        validate: configSchema.validate,
      };
    case "enum":
    case "boolean":
    case "file":
    case "string":
    case "number":
      return {
        out: context.fieldName || "",
        default: configSchema.default,
        validate: configSchema.validate,
      };
  }
}

type Props = {
  config: ConfigSchemaPart,
  onAddPlan(plan: { config: mixed, name: string }): void,
  packages: Array<Package>,
};

export function getFormJobs(value, props) {
  const model = {
    name: {
      out: "name",
      validate: composeValidation(notEmpty(), () => true),
      default: "",
    },
    config: {
      ...convertsConfigSchemaToFormModel(value.config, props, props.config),
      out: "config",
    },
    stager: {
      out: "stager",
      convertOut: v => ({
        expressionValue: v.expressionValue,
        caseValue:
          v.expressionValue === "remote"
            ? {
                ...v.caseValue.host,
                transport: {
                  type: v.caseValue.method.expressionValue,
                  ...v.stager.caseValue.method.caseValue,
                },
              }
            : null,
      }),
      type: {
        out: "expressionValue",
        default: "integrated",
      },
      host: {
        out: "caseValue",
        default: null,
      },
    },
    packageInfos: {
      out: "packageInfos",
      default: null,
      convertOut: (v, props) => props.packages.find(p => p.path == v),
    },
  };
  const jobs = convertConversionModelToConversionJobs(model);
  return jobs;
}

export default compose(
  Connecter,
  withProps(({ onAddPlan }) => ({
    onSubmit(value) {
      onAddPlan(value);
    },
    onValidationFailed() {
      if (global.atom) {
        global.atom.notifications.addWarning("Your plan isn't correct");
      }
    },
  })),
  withState("initial", "setInitial", props => ({
    config: getDefault(props.config),
    packageInfos: props.packages[0] ? props.packages[0].path : null,
  })),
  lifecycle({
    componentWillReceiveProps(nextProps: Props) {
      if (this.props.packages.length === 0 && nextProps.packages.length !== 0) {
        this.props.setInitial({
          ...this.props.initial,
          packageInfos: nextProps.packages[0]
            ? nextProps.packages[0].path
            : null,
        });
      }
    },
  }),
  FormController(
    undefined,
    (value, props) => {
      const v = convertIn(value, getFormJobs(value, props), props);
      return v;
    },
    (value, props) => convertOut(value, getFormJobs(value, props), props),
    { checkIfModified: false },
    (value, props) =>
      composeValidation(
        (value, props) => console.log(value) || true,
        validateModel(getFormJobs(value, props)),
      )(value, props),
  ),
  StateDispatcher(),
)(PlanConfigurer);
