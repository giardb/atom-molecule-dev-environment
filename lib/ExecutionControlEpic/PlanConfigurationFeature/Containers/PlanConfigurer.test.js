import { getFormJobs } from "./PlanConfigurer";

const configSchema = {
  type: "object",
  schemas: {
    configFile: {
      type: "file",
      title: "configfile",
      default: "",
      tester: (packagePath: string, dirname: string) =>
        path.basename(packagePath).indexOf("jest.config") !== -1 ||
        path.basename(packagePath).indexOf(".jest.") !== -1 ||
        path.basename(packagePath) === "package.json",
    },
    modifiedOnly: {
      type: "boolean",
      title: "only modified files (must be in a Git repo)",
      default: false,
    },
    watch: {
      type: "boolean",
      title: "Watch mode",
      default: true,
    },
    binary: {
      type: "conditional",
      expression: {
        type: "enum",
        enum: [
          { value: "local", description: "local" },
          { value: "global", description: "global" },
        ],
      },
      cases: {
        local: null,
        global: {
          type: "object",
          schemas: {
            a: {
              type: "string",
            },
          },
        },
      },
    },
    environmentVariables: {
      type: "array",
      title: "Environment Variables",
      items: {
        type: "string",
        title: "Variable",
        placeholder: "NAME=value",
      },
    },
  },
};

describe("PlanConfigurer", () => {
  describe("getFormJobs", () => {
    it("converts to jobs", () => {
      const subject = getFormJobs({}, { config: configSchema });

      expect(subject).toMatchSnapshot();
    });

    it("converts to jobs taking value in account", () => {
      const subject = getFormJobs(
        {
          environmentVariables: { 0: "HAHA=HEHE", 1: "BABA=BEBE" },
          binary: { expressionValue: "global", caseValue: "" },
        },
        { config: configSchema },
      );

      expect(subject).toMatchSnapshot();
    });
  });
});
