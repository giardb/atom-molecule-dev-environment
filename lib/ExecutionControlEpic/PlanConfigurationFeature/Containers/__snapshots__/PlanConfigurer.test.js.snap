// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanConfigurer getFormJobs converts to jobs 1`] = `
Immutable.List [
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "",
    "in": "name",
    "out": "name",
    "validate": [Function],
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config",
    "out": "config",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "",
    "in": "config.configFile",
    "out": "config.configFile",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": false,
    "in": "config.modifiedOnly",
    "out": "config.modifiedOnly",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": true,
    "in": "config.watch",
    "out": "config.watch",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.binary",
    "out": "config.binary",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.environmentVariables",
    "out": "config.environmentVariables",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "stager",
    "out": "stager",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "integrated",
    "in": "stager.type",
    "out": "stager.expressionValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": null,
    "in": "stager.host",
    "out": "stager.caseValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "packageInfos",
    "out": "packageInfos",
    "validate": undefined,
  },
]
`;

exports[`PlanConfigurer getFormJobs converts to jobs taking value in account 1`] = `
Immutable.List [
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "",
    "in": "name",
    "out": "name",
    "validate": [Function],
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config",
    "out": "config",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "",
    "in": "config.configFile",
    "out": "config.configFile",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": false,
    "in": "config.modifiedOnly",
    "out": "config.modifiedOnly",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": true,
    "in": "config.watch",
    "out": "config.watch",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.binary",
    "out": "config.binary",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.binary.expressionValue",
    "out": "config.binary.expressionValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.binary.caseValue",
    "out": "config.binary.caseValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.binary.caseValue.a",
    "out": "config.binary.caseValue.a",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.environmentVariables",
    "out": "config.environmentVariables",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.environmentVariables.0",
    "out": "config.environmentVariables.0",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "config.environmentVariables.1",
    "out": "config.environmentVariables.1",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "stager",
    "out": "stager",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": "integrated",
    "in": "stager.type",
    "out": "stager.expressionValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": null,
    "in": "stager.host",
    "out": "stager.caseValue",
    "validate": undefined,
  },
  Object {
    "convertIn": [Function],
    "convertOut": [Function],
    "default": undefined,
    "in": "packageInfos",
    "out": "packageInfos",
    "validate": undefined,
  },
]
`;
